pkg/manager/runnable_group.go

```diff
@@ -65,8 +67,25 @@ func (r *runnables) Add(fn Runnable) error {
 		})
 	case webhook.Server:
 		return r.Webhooks.Add(fn, nil)
-	case LeaderElectionRunnable:
-		if !runnable.NeedLeaderElection() {
+	case warmupRunnable, LeaderElectionRunnable:
+		if warmupRunnable, ok := fn.(warmupRunnable); ok {
```

Comment:
If the runnable is a leaderelection runnable and has warmup enabled but leaderelection disabled, we will still do the warmup and then _I think_ the call to warmup might race with the call to `Start` as we start them in parallel? it looks to me like calling `Start` first and then `Warmup` will start the sources twice?

While on paper the above sounds like a nonsensical config it might happen if someone globally enables warmup for all their controllers but has one controller that does not require leader election. Ideally, things just work in that case.

Reply:
What do you think about 
```go
	// Start WarmupRunnables and wait for warmup to complete.
	if err := cm.runnables.Warmup.Start(cm.internalCtx); err != nil {
		return fmt.Errorf("failed to start warmup runnables: %w", err)
	}

	// Start the non-leaderelection Runnables after the cache has synced.
	if err := cm.runnables.Others.Start(cm.internalCtx); err != nil {
		return fmt.Errorf("failed to start other runnables: %w", err)
	}
```

This should solve it, right? As Warmup.Start waits for completion

Reply:
> This should solve it, right? As Warmup.Start waits for completion

Ah you are right, perfect.

This makes me aware of the fact that now that this is not exported anymore we might not need the "wait for finish" logic if we synchronize inside the controller instead, the main reason I asked for that is that if someone outside of c-r implements the interface they would likely expect warmup to finish before being started. So we can do whichever is overall simpler to implement.

Reply:
> we might not need the "wait for finish" logic

If you are referring to the `WaitForWarmupComplete` method, we still need it because it's how we tell the runnable group to make Warmup.Start wait for completion

Reply:
> If I see correctly using sync.Once would just be better in general?

yes. I just didn't know sync.Once existed when I wrote the code 😄

Reply:
Thought about this a bit more. Does ordering it to do warmup before non leader election runnables break backwards compatibility? Previously a manager with non leader election runnables start right after the cache runnables have been started, but now they block on Warmup to complete. 

For controllers this does not matter as they would block on the event source starting anyways, but this breaks in the case where a user has both controllers and some other runnable implemented in user land added to the manager, since now that other runnable has to wait for the controller's source to finish before it can start.

Reply:
Good point. Going back to the original comment

> If the runnable is a leaderelection runnable and has warmup enabled but leaderelection disabled, we will still do the warmup and then I think the call to warmup might race with the call to Start as we start them in parallel? it looks to me like calling Start first and then Warmup will start the sources twice?

If we are using sync.Once we will never start the sources twice.
As we are not exporting the interface anymore Controller will be the only implementation of warmupRunnable. So it seems safe to rely on that?

Reply:
Maybe another way of looking at this is that to preserve backwards compatibility we need the following behavior
- If there is an outstanding call to `startEventSources` then all later calls to the same function should block until that outstanding call returns. I *think* sync.Once already guarantees this, but going to add a UT to verify this behavior. 
- Any function that calls `startEventSources` needs to hold `c.mu` because the critical section on L268 modifies `startWatches` outside of the startEventSources call. 

Simplest solution is probably to take the mu lock in the warmup call.

Reply:
Just realized that I don't know if sync.Once blocks additional callers

Reply:
> Any function that calls startEventSources needs to hold c.mu because the critical section on L268 modifies startWatches outside of the startEventSources call.

Maybe worth checking if l.268 should be moved into `startEventSources` (not sure if it changes the fact that we have to hold the lock, but it might be a good improvement in any case)

Reply:
> Just realized that I don't know if sync.Once blocks additional callers until the code within sync.Once is completed

Looks like it does, so that solves 1 problem: https://go.dev/play/p/XAFFCESpUOe

Reply:
Pushed code with tests regarding the race + clearing out `startWatches`.